### PR TITLE
DOC: timezone warning for DST beyond 2038-01-18

### LINKED ISSUE #33061: DateTimeIndex.tz_convert() does not apply DST from 2038 onward
#### Code Sample, a copy-pastable example if possible

```python
assert np.all(pd.date_range('1/1/2037', periods=8760, freq='H', tz='EST').time == pd.date_range('1/1/2037', periods=8760, freq='H', tz='US/Eastern').time) == False

assert np.all(pd.date_range('1/1/2038', periods=8760, freq='H', tz='EST').time == pd.date_range('1/1/2038', periods=8760, freq='H', tz='US/Eastern').time) == False # fails

assert np.all(pd.date_range('1/1/2039', periods=8760, freq='H', tz='EST').time == pd.date_range('1/1/2039', periods=8760, freq='H', tz='US/Eastern').time) == False # fails

```
#### Problem description

Wow, this one hurt. US/Eastern timezone is DST-adjusted (blend of EST/EDT) whereas EST is just EST. 

The second and third assert statements above should both return False.

Surprised this hasn't come up before. 

This is apparently related to a UNIX issue: https://en.wikipedia.org/wiki/Year_2038_problem. With that said, it seems the dtype is datetime64 with some pandas customizations on timezone. Supposedly 64 bit should have solved this.

#### Expected Output

Both of the following should pass:
assert np.all(pd.date_range('1/1/2038', periods=8760, freq='H', tz='EST').time == pd.date_range('1/1/2038', periods=8760, freq='H', tz='US/Eastern').time) == False # fails

assert np.all(pd.date_range('1/1/2039', periods=8760, freq='H', tz='EST').time == pd.date_range('1/1/2039', periods=8760, freq='H', tz='US/Eastern').time) == False # fails

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
NSTALLED VERSIONS
------------------
commit           : None
python           : 3.8.1.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 142 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252

pandas           : 1.0.0
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 45.1.0.post20200127
Cython           : None
pytest           : 5.3.5
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.1
IPython          : 7.11.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : 2.7.1
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 0.13.0
pytables         : None
pytest           : 5.3.5
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : 3.6.1
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
xlsxwriter       : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by telferm57**
Add warning to time_series user guide that after 2038-01-18 DST will not be respected in tz aware dates 

- [x ] closes #33061
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by pganssle** [comment]
> If you are using dates beyond 18 Jan 2038, note that pandas does not apply daylight saving time adjustments to timezone aware dates. This is partly because the underlying libraries do not currently address the Year 2038 Problem, and partly because there is some discussion on how reliable any DST settings that far into the future will be.

As mentioned in the issue, I do not think this wording is accurate. `pandas` only ever uses the UTC offsets provided by the libraries. Right now the libraries that `pandas` supports do not have any transitions after 2038-01-18.

I think if you're going to warn against this, something more circumspect would be a good idea, like:

> Time zone data for far future time zones are likely to be inaccurate, as they are simple extrapolations of the current set of rules. In general, it is inadvisable to rely on the accuracy of time zone data for far future datetimes. This is especially true for datetimes after 18 January 2038, since some time zone libraries do not support transitions after that point due to the Year 2038 Problem.

I'm vague about the "some time zone libraries" since I know that `dateutil` is planning to add support soon-ish and the maintainer of `pytz` has said that he's intending to turn `pytz` into a thin wrapper around PEP 615 when it is available (and the PEP 615 `zoneinfo` module *does* support these), so it is likely that any strong wording here will become inaccurate soon.

File: doc/source/user_guide/timeseries.rst
Comment (mroeschke) [inline doc/source/user_guide/timeseries.rst:2270]: NIT: Could we ISO format the date? (`dates beyond 2038-01-18`)

**Comment by telferm57** [comment]
Thanks for the response. I think we need to get the 2038 issue in there first, because that is the symptom users will see. Also, I didn't want to go into any details of the cause of version 1 libraries being used to store transition dates in signed 32-bit fields, as I wanted to keep it quite short - there are already 3 other warnings in that section of the guide!  I don't think people are that interested in where the cause of the issue is, as long as it will be fixed well before the epochalypse! (love that term!  probably reveals my tz naivety that I haven't heard it before) . 

Proposed new wording: 

If you are using dates beyond 2038-01-18, due to current deficiencies in the underlying libraries caused by the year 2038 problem, daylight saving time (DST) adjustments to timezone aware dates will not be applied. The underlying libraries will be fixed at some point, which will also correct this behaviour. It should be noted though, that time zone data for far future time zones are likely to be inaccurate, as they are simple extrapolations of the current set of (regularly revised) rules.

**Comment by mroeschke** [comment]
> The underlying libraries will be fixed at some point.

I think this guarantee is too strong. I think the phrasing should just imply that the fix is on the timezone library side.

**Comment by telferm57** [comment]
If and when the underlying libraries are fixed, the DST transitions will be applied?

**Comment by mroeschke** [comment]
Sure. If you could push up the revised changes, it will be easier to review inline from the changes than here in the comments.

File: doc/source/user_guide/timeseries.rst
Comment (joooeey) [inline doc/source/user_guide/timeseries.rst:2270]: The second part `and partly because there is some discussion on how reliable any DST settings that far into the future will be` sounds vague. There are not just discussions. It seems arbitrary to start raising this concern in 2038. With so many jurisdictions in the world, some will change their time zones earlier. For instance, the EU is on track to abolish DST in the 2020s. I think this sentence should be left out because it's a separate issue.

Or that half-sentence could be put in a separate warning but I think that would be overkill since this issue is not Pandas-specific. Our docs will be very cluttered if we warn about every generic issue that might occur with date and time handling.

Comment (telferm57) [reply doc/source/user_guide/timeseries.rst:2270]: IIUC it is only DST transitions that are affected , and only after 2038. So any changes up to 2038 will be reflected correctly in construction of a tz aware time ... have I understood you OK? The exact sentence you quoted has been removed.

**Review by mroeschke — APPROVED** [review]

File: doc/source/user_guide/timeseries.rst
Comment (joooeey) [reply doc/source/user_guide/timeseries.rst]: No, changes will not be reflected correctly until 2038 because you forgot to `import crystalball` (and even if you had imported it you'd find it can't be used to read the future). However, I think **the warning should only address the 2038 issue**. One could make an extra warning for the general issue of real-world timezones being unpredictable in the future but I don't think that's necessary because that's what people expect.

The new sentence `It should be noted though, that time zone data for far future time zones are likely to be inaccurate, as they are simple extrapolations of the current set of (regularly revised) rules.` is misleading and confusing. One can't even predict timezone switches a day in advance, as we have seen [in 2018 in Morocco](https://www.bbc.com/news/world-africa-45995634), let alone 18 years.

Also, the example you give is only about the 2038 problem. Especially in Britain it's likely that they abolish daylight saving together with the EU before 2038. The quoted sentence just makes it a little bit harder for people to follow that example.

Comment (telferm57) [reply doc/source/user_guide/timeseries.rst]: yes, it is quite possible that UK will  stick to permanent DST some time in the next 5 years, but if that happens, the underlying libraries as they are will support that - the changes will be made in pre-2038 dates.  I can't see what the problem is!

Comment (pganssle) [reply doc/source/user_guide/timeseries.rst]: > yes, it is quite possible that UK will stick to permanent DST some time in the next 5 years, but if that happens, the underlying libraries as they are will support that - the changes will be made in pre-2038 dates.

The same argument can be made for post-2038 dates. Certainly the underlying libraries will be fixed before 2038, so the only time this would ever come up is if you're trying to convert a local time in the far future into a timestamp (either in UTC or find out what offset applies), which is basically something that *cannot be known to be accurate*.

This is my problem with the whole idea of adding this warning ­– it's saying, "If you are trying to do this thing you shouldn't do, the answer might be different from what you expect." It may be that the zone you're in has eliminated DST by then, in which case the answer is *right* and the "correct" rule is the one that's wrong. It's not a problem particular to `pandas`, and it's not easy to convey to end users what the problem is and what should be done about it, so a warning in the documentation doesn't seem like a good fit.

I still don't like the focus on 2038 as the cut-off point, because it makes it seem like adding version 2 support to the underlying libraries will fix the problem, but in reality the problem is that the users are trying to do this *at all*, and they just notice something out of the ordinary after 2038.

File: doc/source/user_guide/timeseries.rst
Comment (pganssle) [inline doc/source/user_guide/timeseries.rst:2282]: Why is this called `DST` instead of `LON` or some other thing?

File: doc/source/user_guide/timeseries.rst
Comment (pganssle) [inline doc/source/user_guide/timeseries.rst:2283]: This does not seem right to me.

If you're going to keep the example (and I'm not convinced you should, *especially* since it's right around when a DST transition happens - if anything you should move it deep into summer to guarantee that the fluctuations aren't just due to the transition moving around), it would be better to make assertions about the thing you care about.

```python
assert pd.Timestamp(d_2037, tz=LON).tzname() != "GMT"
assert pd.Timestamp(d_2038, tz=LON).tzname() != "GMT"
```

Even better, though, would be a repr:

```python
>>> pd.Timestamp(d_2037, tz=LON).tzname()
'BST'
>>> pd.Timestamp(d_2037, tz=LON).tzname()
'GMT'
```

File: doc/source/user_guide/timeseries.rst
Comment (joooeey) [reply doc/source/user_guide/timeseries.rst]: @pganssle you're essentially saying that people shouldn't use timezones on any far future dates that they deal with. (side note: I disagree with the word far, the future is unpredictable, far or near.) This is unrealistic. While many applications could rely on UT, UTC, or system time, I'm sure you can come up with many use cases where developers need to deal with time zones in future dates (e.g. when opening hours, TV schedules or flight schedules are involved).

Obviously, nobody can guarantee that future time zones will be _correct_. However, developers expect a _predictable_ behaviour of the library. Now, the 2038 issue breaks that _predictability_. It's common sense that the future cannot be predicted, so there is no need to warn about potential future political changes. However, developers should be warned if code yields unexpected results. 

That's why we should warn developers about the 2038 issue (as in the PR title). And the following sentence should be removed because it has nothing to do with the 2038 issue: `It should be noted though, that time zone data for far future time zones are likely to be inaccurate, as they are simple extrapolations of the current set of (regularly revised) rules.`

@pganssle if you think we need to warn developers about using timezones when dealing with future dates, this should go in a separate pull request.

File: doc/source/user_guide/timeseries.rst
Comment (telferm57) [reply doc/source/user_guide/timeseries.rst:2282]: because I am focusing on DST transitions - just happen to have picked London as that's local

File: doc/source/user_guide/timeseries.rst
Comment (telferm57) [reply doc/source/user_guide/timeseries.rst:2283]: I agree summer date would be clearer , no so sure whether your example is clearer , or whether using BST (local zone name for DST) is clearer for a global audience ... hmmm

Comment (pganssle) [reply doc/source/user_guide/timeseries.rst:2283]: It doesn't really matter what offsets you use there, pick anything. The important thing is that it's obvious that they are *different* on the same date in different years, and that it's obvious that that's not due to fluctuations one way or the other in the date of the DST change.

File: doc/source/user_guide/timeseries.rst
Comment (pganssle) [reply doc/source/user_guide/timeseries.rst:2282]: People will think it means "this is the Daylight Saving Time zone". You don't have to call it `LON`, but you should at least call it `ZONE` or something.

File: doc/source/user_guide/timeseries.rst
Comment (pganssle) [reply doc/source/user_guide/timeseries.rst]: > @pganssle you're essentially saying that people shouldn't use timezones on any far future dates that they deal with. (side note: I disagree with the word far, the future is unpredictable, far or near.) This is unrealistic. While many applications could rely on UT, UTC, or system time, I'm sure you can come up with many use cases where developers need to deal with time zones in future dates (e.g. when opening hours, TV schedules or flight schedules are involved).

No, I'm saying that you should *only* use "time zones" in future dates and that conversion to UTC is increasingly unreliable the further into the future you go. 18 years is a long way into the future, so this is like saying, "Don't forget to bring a bathing suit if you jump into shark-infested water!"

It's somewhat misleading to include a warning like this without explaining that yes it's not what you might expect but it's *basically* not a problem at the moment, because if you are relying on accuracy in this situation you have bigger problems. It also implies that dates in December 2037 can be used somewhat accurately.

Comment (joooeey) [reply doc/source/user_guide/timeseries.rst]: Okay. I'm convinced now. We should warn about converting (any) future time. I still think this warning shouldn't be an afterthought in the 2038 warning text. It should be its own warning box. That gives it the prominence it deserves and ensures the warning remains if the 2038 problem is resolved in the downstream libraries and that box is removed.

Something along the lines of:

> Be aware that for times in the future, correct conversion between time zones (and UTC) cannot be guaranteed by any time zone library. Sometimes the rules governing a timezone's offset from UTC are changed. Authorities usually announce such changes many months in advance but there have been examples of much shorter lead times such as when Morocco announced just two days before the planned switch from summer time to winter time in 2018 that the country would stay on summer time permanently. Furthermore, the databases that Pandas relies on may need some time to record planned changes to timezone offsets.

File: doc/source/user_guide/timeseries.rst
Comment (telferm57) [reply doc/source/user_guide/timeseries.rst:2283]: Is it OK to submit further changes once the PR has been approved?

Comment (mroeschke) [reply doc/source/user_guide/timeseries.rst:2283]: Yes you can submit further changes. I think @pganssle's suggestion about using the repr would be clearer.

**Comment by jreback** [comment]
thanks @telferm57 

happy to have clarifications in another PR. but this looks good for now.

File: doc/source/user_guide/timeseries.rst
Comment (joooeey) [reply doc/source/user_guide/timeseries.rst]: I made a PR to add this text: #34100

**Pull request merged**
